Panic on unexpected value passed to validation

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers .
-->

### Terraform Version
<!---
Run `terraform -v` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
Terraform v0.12.5
```

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->

```
panic: reflect: call of reflect.Value.Type on zero Value
2019-07-26T11:52:48.026+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: 
2019-07-26T11:52:48.026+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: goroutine 66 [running]:
2019-07-26T11:52:48.026+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: reflect.Value.Type(0x0, 0x0, 0x0, 0x38, 0x0)
2019-07-26T11:52:48.026+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/goenv/versions/1.12.6/src/reflect/value.go:1813 +0x169
2019-07-26T11:52:48.026+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.schemaMap.validatePrimitive(0xc00058d4a0, 0xc000cc1f20, 0x18, 0x0, 0x0, 0xc00015cc00, 0xc000ae3350, 0x1, 0x3d0c6e0, 0xc000c50960, ...)
2019-07-26T11:52:48.026+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/schema.go:1694 +0xc8
2019-07-26T11:52:48.026+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.schemaMap.validateType(0xc00058d4a0, 0xc000cc1f20, 0x18, 0x0, 0x0, 0xc00015cc00, 0xc000ae3350, 0x4088cc, 0xc00001e000, 0x4394cc0, ...)
2019-07-26T11:52:48.026+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/schema.go:1779 +0x56f
2019-07-26T11:52:48.026+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.schemaMap.validateList(0xc00058d4a0, 0x4baf375, 0x16, 0x3d0c6e0, 0xc000c50960, 0xc00015ef00, 0xc000ae3350, 0x0, 0x41a7a00, 0xc000ae3380, ...)
2019-07-26T11:52:48.026+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/schema.go:1495 +0x86c
2019-07-26T11:52:48.026+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.schemaMap.validateType(0xc00058d4a0, 0x4baf375, 0x16, 0x3d0c6e0, 0xc000c50960, 0xc00015ef00, 0xc000ae3350, 0x0, 0x0, 0x0, ...)
2019-07-26T11:52:48.026+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/schema.go:1775 +0x9e
2019-07-26T11:52:48.026+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.schemaMap.validate(0xc00058d4a0, 0x4baf375, 0x16, 0xc00015ef00, 0xc000ae3350, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/schema.go:1382 +0x213
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.schemaMap.validateObject(0xc00058d4a0, 0x0, 0x0, 0xc00058d4a0, 0xc000ae3350, 0x4b8d20d, 0xc000cb8b10, 0xf, 0xc000510601, 0x7a0000c000c509e0, ...)
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/schema.go:1661 +0x1c6
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.schemaMap.Validate(...)
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/schema.go:677
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Validate(0xc0004e2b80, 0xc000ae3350, 0xc000cb8b10, 0xf, 0xc0002343c8, 0x1, 0x40, 0x0)
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:357 +0x5e
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).ValidateResource(0xc000230280, 0xc000cb8b10, 0xf, 0xc000ae3350, 0xc000b3ff40, 0xc000ae3350, 0xc00011f7e8, 0x41a7a00, 0xc000ae32f0, 0x0)
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:242 +0x1d8
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/plugin.(*GRPCProviderServer).ValidateResourceTypeConfig(0xc00011ef68, 0x5419c40, 0xc0004fd3e0, 0xc00087eec0, 0xc00011ef68, 0xc0004fd3e0, 0xc0006a1bd0)
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/plugin/grpc_provider.go:206 +0x1d9
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/internal/tfplugin5._Provider_ValidateResourceTypeConfig_Handler(0x4a66240, 0xc00011ef68, 0x5419c40, 0xc0004fd3e0, 0xc000c85680, 0x0, 0x5419c40, 0xc0004fd3e0, 0xc000c92500, 0x4bf)
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/internal/tfplugin5/tfplugin5.pb.go:3109 +0x23e
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001cf800, 0x543b580, 0xc0005a6900, 0xc000012b00, 0xc00059ee10, 0x925da70, 0x0, 0x0, 0x0)
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/google.golang.org/grpc/server.go:966 +0x470
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/google.golang.org/grpc.(*Server).handleStream(0xc0001cf800, 0x543b580, 0xc0005a6900, 0xc000012b00, 0x0)
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/google.golang.org/grpc/server.go:1245 +0xd25
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc00003a4d0, 0xc0001cf800, 0x543b580, 0xc0005a6900, 0xc000012b00)
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/google.golang.org/grpc/server.go:685 +0x9f
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4: created by github.com/terraform-providers/terraform-provider-aws/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
2019-07-26T11:52:48.027+0200 [DEBUG] plugin.terraform-provider-aws_v2.21.0_x4:  /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/google.golang.org/grpc/server.go:683 +0xa1
2019-07-26T11:52:48.030+0200 [DEBUG] plugin: plugin process exited: path=/home/mtesch/Documents/Kunden/Allianz/a2l-infra-shared-modules/rds/.terraform/plugins/linux_amd64/terraform-provider-aws_v2.21.0_x4 pid=28714 error="exit status 2"
2019/07/26 11:52:48 [ERROR] <root>: eval: *terraform.EvalDiff, err: rpc error: code = Unavailable desc = transport is closing
2019/07/26 11:52:48 [ERROR] <root>: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2019/07/26 11:52:48 [TRACE] [walkPlan] Exiting eval tree: aws_db_instance.db_instance
2019/07/26 11:52:48 [TRACE] vertex "aws_db_instance.db_instance": visit complete
2019/07/26 11:52:48 [TRACE] vertex "aws_db_instance.db_instance": dynamic subgraph encountered errors
2019/07/26 11:52:48 [TRACE] vertex "aws_db_instance.db_instance": visit complete
2019/07/26 11:52:48 [ERROR] <root>: eval: *terraform.EvalDiff, err: rpc error: code = Unavailable desc = transport is closing
2019/07/26 11:52:48 [ERROR] <root>: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2019/07/26 11:52:48 [TRACE] [walkPlan] Exiting eval tree: aws_kms_alias.db_kms_key_alias[0]
2019/07/26 11:52:48 [TRACE] dag/walk: upstream of "output.db_instance_endpoint" errored, so skipping
2019/07/26 11:52:48 [TRACE] vertex "aws_kms_alias.db_kms_key_alias[0]": visit complete
2019/07/26 11:52:48 [TRACE] vertex "aws_kms_alias.db_kms_key_alias": dynamic subgraph encountered errors
2019/07/26 11:52:48 [TRACE] vertex "aws_kms_alias.db_kms_key_alias": visit complete
2019/07/26 11:52:48 [TRACE] dag/walk: upstream of "meta.count-boundary (EachMode fixup)" errored, so skipping
2019/07/26 11:52:48 [TRACE] dag/walk: upstream of "provider.aws (close)" errored, so skipping
2019/07/26 11:52:48 [TRACE] dag/walk: upstream of "root" errored, so skipping
2019/07/26 11:52:48 [INFO] backend/local: plan operation completed
2019/07/26 11:52:48 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info
2019/07/26 11:52:48 [TRACE] statemgr.Filesystem: unlocking terraform.tfstate using fcntl flock
2019-07-26T11:52:48.031+0200 [DEBUG] plugin: plugin exited
``` 

### Expected Behavior
terraform plan for aws_db_instance should succeed.

### Actual Behavior
terraform crashes

### Steps to Reproduce
terraform plan with a "aws_db_instance" resource

--- a/helper/schema/schema.go
+++ b/helper/schema/schema.go
@@ -1440,6 +1440,11 @@ func (m schemaMap) validateList(
 		}
 	}
 
+	// schemaMap can't validate nil
+	if raw == nil {
+		return nil, nil
+	}
+
 	// We use reflection to verify the slice because you can't
 	// case to []interface{} unless the slice is exactly that type.
 	rawV := reflect.ValueOf(raw)
@@ -1518,6 +1523,10 @@ func (m schemaMap) validateMap(
 		}
 	}
 
+	// schemaMap can't validate nil
+	if raw == nil {
+		return nil, nil
+	}
 	// We use reflection to verify the slice because you can't
 	// case to []interface{} unless the slice is exactly that type.
 	rawV := reflect.ValueOf(raw)
@@ -1644,6 +1653,12 @@ func (m schemaMap) validateObject(
 	schema map[string]*Schema,
 	c *terraform.ResourceConfig) ([]string, []error) {
 	raw, _ := c.Get(k)
+
+	// schemaMap can't validate nil
+	if raw == nil {
+		return nil, nil
+	}
+
 	if _, ok := raw.(map[string]interface{}); !ok && !c.IsComputed(k) {
 		return nil, []error{fmt.Errorf(
 			"%s: expected object, got %s",
@@ -1688,6 +1703,14 @@ func (m schemaMap) validatePrimitive(
 	raw interface{},
 	schema *Schema,
 	c *terraform.ResourceConfig) ([]string, []error) {
+
+	// a nil value shouldn't happen in the old protocol, and in the new
+	// protocol the types have already been validated. Either way, we can't
+	// reflect on nil, so don't panic.
+	if raw == nil {
+		return nil, nil
+	}
+
 	// Catch if the user gave a complex type where a primitive was
 	// expected, so we can return a friendly error message that
 	// doesn't contain Go type system terminology.

--- a/helper/schema/schema_test.go
+++ b/helper/schema/schema_test.go
@@ -5465,6 +5465,39 @@ func TestSchemaMap_Validate(t *testing.T) {
 			},
 			Err: false,
 		},
+
+		"unexpected nils values": {
+			Schema: map[string]*Schema{
+				"strings": &Schema{
+					Type:     TypeList,
+					Optional: true,
+					Elem: &Schema{
+						Type: TypeString,
+					},
+				},
+				"block": &Schema{
+					Type:     TypeList,
+					Optional: true,
+					Elem: &Resource{
+						Schema: map[string]*Schema{
+							"int": &Schema{
+								Type:     TypeInt,
+								Required: true,
+							},
+						},
+					},
+				},
+			},
+
+			Config: map[string]interface{}{
+				"strings": []interface{}{"1", nil},
+				"block": []interface{}{map[string]interface{}{
+					"int": nil,
+				},
+					nil,
+				},
+			},
+		},
 	}
 
 	for tn, tc := range cases {

